libraries/lib-cloud-audiocom/UploadService.cpp

```diff
@@ -0,0 +1,475 @@
+/*  SPDX-License-Identifier: GPL-2.0-or-later */
+/*!********************************************************************
+
+  Audacity: A Digital Audio Editor
+
+  OAuthService.cpp
+
+  Dmitry Vedenko
+
+**********************************************************************/
+
+#include "UploadService.h"
+
+#include <mutex>
+
+#include <wx/filefn.h>
+#include <wx/filename.h>
+
+#include <rapidjson/document.h>
+#include <rapidjson/writer.h>
+
+#include "OAuthService.h"
+#include "ServiceConfig.h"
+
+#include "NetworkManager.h"
+#include "Request.h"
+#include "IResponse.h"
+#include "MultipartData.h"
+
+#include "CodeConversions.h"
+
+namespace cloud::audiocom
+{
+namespace
+{
+std::string_view DeduceMimeType(const wxString& ext)
+{
+   if (ext == "wv")
+      return "audio/x-wavpack";
+   else if (ext == "flac")
+      return "audio/x-flac";
+   else if (ext == "mp3")
+      return "audio/mpeg";
+   else
+      return "audio/x-wav";
+}
+
+std::string GetUploadRequestPayload(const wxString& filePath, const wxString& projectName)
+{
+   rapidjson::Document document;
+   document.SetObject();
+
+   const wxFileName fileName(filePath);
+   const auto mimeType = DeduceMimeType(fileName.GetExt());
+
+   document.AddMember(
+      "mime",
+      rapidjson::Value(
+         mimeType.data(), mimeType.length(), document.GetAllocator()),
+      document.GetAllocator());
+
+   const auto name = audacity::ToUTF8(projectName.empty() ? fileName.GetFullName() : projectName);
+
+   document.AddMember(
+      "name",
+      rapidjson::Value(name.data(), name.length(), document.GetAllocator()),
+      document.GetAllocator());
+
+   document.AddMember(
+      "size",
+      rapidjson::Value(static_cast<int64_t>(fileName.GetSize().GetValue())),
+      document.GetAllocator());
+
+   rapidjson::StringBuffer buffer;
+   rapidjson::Writer<rapidjson::StringBuffer> writer(buffer);
+   document.Accept(writer);
+
+   return std::string(buffer.GetString());
+}
+
+std::string GetProgressPayload(uint64_t current, uint64_t total)
+{
+   rapidjson::Document document;
+   document.SetObject();
+
+   document.AddMember(
+      "progress", rapidjson::Value(current / static_cast<double>(total) * 100.0),
+      document.GetAllocator());
+
+   rapidjson::StringBuffer buffer;
+   rapidjson::Writer<rapidjson::StringBuffer> writer(buffer);
+   document.Accept(writer);
+
+   return std::string(buffer.GetString());
+}
+
+
+// This class will capture itself inside the request handlers
+// by a strong reference. This way we ensure that it outlives all
+// the outstanding requests.
```

Comment:
... so instead, the destructor of the concrete class could guarantee that one of completion, abort, or discard has happened, and then the lambdas should capture weak, not strong, references to self.

I think this would make better RAII guarantees on the main thread.  The "leak" in case there is no explicit abort on the main thread isn't until program exit, but it could mean the other thread works longer than needed before stopping.

Reply:
This will complicate the code. Abort is not synchronous and I don't see much reason to change that (although it is achievable). Passing weak pointers will additionally increase complexity of the implementation.

Right now Promise (or future, or operation) outlives the dialog at most until lubcurl invokes any of it's callbacks. Making Abort synchronous would basically just delay the dialog closing for that specific time. I do not, however, like the idea for blocking UI thread for such reasons.

Reply:
Did I misunderstand the intended context of use of this, then?

Is there going to be a non-modal dialog in Audacity giving an interface that lets the user abort the operation?  Or will that be on a web page only?

Reply:
There is a modal dialog in Audacity that calls Abort when closing, if the promise was not fulfilled.